File: packages/npm/scripts/stage.mjs
#!/usr/bin/env node
//
// Stages WASM build artefacts into packages/npm/dist/ for publication.
//
// Inputs (defaults; override via --build-dir / --out-dir):
//   --build-dir build-wasm
//   --out-dir   packages/npm/dist
//
// Copies:
//   <build-dir>/formulon.js   -> <out-dir>/formulon_core.js
//   <build-dir>/formulon.wasm -> <out-dir>/formulon.wasm
//   packages/npm/index.mjs     -> <out-dir>/formulon.js
//   src/wasm/formulon.d.ts    -> <out-dir>/formulon.d.ts
//
// Run via `make npm-package`. No npm dependencies; Node 18 stdlib only.

import { constants as FS } from 'node:fs';
import { access, copyFile, mkdir } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// Repo root is two levels above this script (packages/npm/scripts/).
const repoRoot = path.resolve(__dirname, '..', '..', '..');

function parseArgs(argv) {
  const out = { buildDir: 'build-wasm', outDir: 'packages/npm/dist' };
  for (let i = 0; i < argv.length; ++i) {
    const a = argv[i];
    if (a === '--build-dir') {
      out.buildDir = argv[++i];
    } else if (a === '--out-dir') {
      out.outDir = argv[++i];
    } else if (a === '-h' || a === '--help') {
      console.log('Usage: stage.mjs [--build-dir <path>] [--out-dir <path>]');
      process.exit(0);
    } else {
      console.error(`stage.mjs: unknown argument: ${a}`);
      process.exit(2);
    }
  }
  return out;
}

async function fileExists(p) {
  try {
    await access(p, FS.R_OK);
    return true;
  } catch {
    return false;
  }
}

async function main() {
  const { buildDir, outDir } = parseArgs(process.argv.slice(2));
  const absBuildDir = path.resolve(repoRoot, buildDir);
  const absOutDir = path.resolve(repoRoot, outDir);

  const jsSrc = path.join(absBuildDir, 'formulon.js');
  const wasmSrc = path.join(absBuildDir, 'formulon.wasm');
  const dtsSrc = path.join(repoRoot, 'src', 'wasm', 'formulon.d.ts');
  const shimSrc = path.join(repoRoot, 'packages', 'npm', 'index.mjs');

  for (const p of [jsSrc, wasmSrc, dtsSrc, shimSrc]) {
    if (!(await fileExists(p))) {
      console.error(`stage.mjs: missing input ${p}`);
      console.error('  Run `make wasm` first to produce the WASM artefacts.');
      process.exit(1);
    }
  }

  await mkdir(absOutDir, { recursive: true });

  await copyFile(jsSrc, path.join(absOutDir, 'formulon_core.js'));
  await copyFile(wasmSrc, path.join(absOutDir, 'formulon.wasm'));
  await copyFile(dtsSrc, path.join(absOutDir, 'formulon.d.ts'));
  await copyFile(shimSrc, path.join(absOutDir, 'formulon.js'));

  console.log(`staged 4 file(s) -> ${absOutDir}`);
}

main().catch((e) => {
  console.error('stage.mjs: fatal:', e && e.stack ? e.stack : e);
  process.exit(1);
});

File: packages/python/scripts/stage.py
#!/usr/bin/env python3
# Copyright 2026 libraz. Licensed under the Apache License, Version 2.0.
"""Stage libformulon into the Python package's _lib/ directory.

Run via ``make python-package``. Stdlib only -- no third-party deps so
this works in any clean Python 3.9+ environment.

Inputs (defaults; override via command-line flags):
  --build-dir build-py
  --config    Release

Steps:
  1. Configure CMake with FM_BUILD_C_API_SHARED=ON.
  2. Build the ``formulon`` shared-library target.
  3. Copy the resulting libformulon.{dylib,so,dll} into
     packages/python/formulon/_lib/.
"""

from __future__ import annotations

import argparse
import platform
import shutil
import subprocess
import sys
from pathlib import Path

# packages/python/scripts/stage.py -> repo root is three levels up.
REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent
PKG_LIB_DIR = REPO_ROOT / "packages" / "python" / "formulon" / "_lib"


def _parse_args(argv: list[str]) -> argparse.Namespace:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument(
        "--build-dir",
        default="build-py",
        help="CMake build directory (relative to repo root). Default: build-py",
    )
    p.add_argument(
        "--config",
        default="Release",
        choices=["Release", "Debug", "RelWithDebInfo", "MinSizeRel"],
        help="CMake build configuration. Default: Release",
    )
    return p.parse_args(argv)


def _run(cmd: list[str], cwd: Path | None = None) -> None:
    """Run ``cmd`` and abort with a clear message on failure."""
    print(f"+ {' '.join(cmd)}", flush=True)
    result = subprocess.run(cmd, cwd=cwd, check=False)
    if result.returncode != 0:
        print(
            f"stage.py: command failed (exit {result.returncode}): {' '.join(cmd)}",
            file=sys.stderr,
        )
        sys.exit(result.returncode)


def _resolve_built_lib(build_dir: Path, config: str) -> Path:
    """Locate the freshly built shared library inside ``build_dir``.

    Different generators put the artefact in different locations:
      * Single-config (Make / Ninja, default on macOS / Linux):
        ``build_dir/libformulon.{dylib,so}``.
      * Multi-config (Visual Studio, Xcode):
        ``build_dir/<Config>/formulon.dll`` (Windows) or
        ``build_dir/<Config>/libformulon.{dylib,so}``.
    """
    system = platform.system()
    if system == "Darwin":
        names = ["libformulon.dylib"]
    elif system == "Windows":
        names = ["formulon.dll", "libformulon.dll"]
    else:
        names = ["libformulon.so"]

    candidates: list[Path] = []
    for name in names:
        candidates.append(build_dir / name)
        candidates.append(build_dir / config / name)

    for c in candidates:
        if c.is_file():
            return c

    raise FileNotFoundError(
        f"stage.py: could not locate built shared library. Looked in:\n"
        + "\n".join(f"  {c}" for c in candidates)
    )


def _windows_import_libs(build_dir: Path, config: str) -> list[Path]:
    """On Windows, the .lib import library should ship next to the .dll."""
    libs: list[Path] = []
    for sub in (build_dir, build_dir / config):
        for name in ("formulon.lib", "libformulon.lib"):
            p = sub / name
            if p.is_file():
                libs.append(p)
    return libs


def main(argv: list[str]) -> int:
    args = _parse_args(argv)
    build_dir = (REPO_ROOT / args.build_dir).resolve()

    # 1. Configure.
    _run(
        [
            "cmake",
            "-B",
            str(build_dir),
            "-S",
            str(REPO_ROOT),
            "-DFM_BUILD_C_API_SHARED=ON",
            "-DFM_BUILD_TESTING=OFF",
            "-DFM_BUILD_CLI=OFF",
            f"-DCMAKE_BUILD_TYPE={args.config}",
        ]
    )

    # 2. Build the shared library only.
    _run(
        [
            "cmake",
            "--build",
            str(build_dir),
            "--target",
            "formulon",
            "--config",
            args.config,
            "--parallel",
        ]
    )

    # 3. Copy artefacts into _lib/.
    PKG_LIB_DIR.mkdir(parents=True, exist_ok=True)

    src = _resolve_built_lib(build_dir, args.config)
    dst = PKG_LIB_DIR / src.name
    shutil.copy2(src, dst)
    print(f"staged libformulon -> {dst} ({dst.stat().st_size} bytes)")

    if platform.system() == "Windows":
        for lib in _windows_import_libs(build_dir, args.config):
            shutil.copy2(lib, PKG_LIB_DIR / lib.name)
            print(f"staged {lib.name} -> {PKG_LIB_DIR / lib.name}")

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
